movici_simulation_core/storage/sqlite_schema.py

```diff
@@ -0,0 +1,329 @@
+        return np.frombuffer(self.data, dtype=self.dtype).reshape(shape)
+
+
+class Update(Base):
+    """Represents a simulation update at a specific timestamp and iteration."""
+
+    __tablename__ = "update"
+    __table_args__ = (
+        UniqueConstraint("timestamp", "iteration", "dataset_name", name="uq_update"),
+        Index("idx_update_time", "timestamp", "iteration"),
+        Index("idx_update_dataset", "dataset_name"),
+    )
+
+    id = Column(Integer, primary_key=True, autoincrement=True)
+    timestamp = Column(Integer, nullable=False)
+    iteration = Column(Integer, nullable=False)
+    dataset_name = Column(String, nullable=False)
+    origin = Column(String)  # Optional: model that produced this update
+
+    # Relationships
+    attributes = relationship(
+        "AttributeData", secondary="update_attribute", back_populates="updates"
+    )
+
+
+class AttributeData(Base):
+    """Stores entity attribute data with support for both uniform and CSR sparse arrays."""
+
+    __tablename__ = "attribute_data"
+    __table_args__ = (
+        Index("idx_attr_entity_group", "entity_group"),
+        Index("idx_attr_name", "attribute_name"),
+    )
+
+    id = Column(Integer, primary_key=True, autoincrement=True)
+    entity_group = Column(String, nullable=False)
+    attribute_name = Column(String, nullable=False)
+
+    # Foreign keys to numpy arrays
+    data_id = Column(Integer, ForeignKey("numpy_array.id", ondelete="CASCADE"), nullable=False)
+    indptr_id = Column(Integer, ForeignKey("numpy_array.id", ondelete="CASCADE"))
+
+    # Optional metadata for optimization
+    min_val = Column(Float)
+    max_val = Column(Float)
+
+    # Relationships
+    data = relationship("NumpyArray", foreign_keys=[data_id])
+    indptr = relationship("NumpyArray", foreign_keys=[indptr_id])
+    updates = relationship("Update", secondary="update_attribute", back_populates="attributes")
+
+    @property
+    def is_sparse(self) -> bool:
+        """Check if this attribute uses CSR sparse representation.
+
+        :return: True if sparse, False otherwise
+        """
+        return self.indptr_id is not None
+
+    def get_data(self) -> dict:
+        """Get attribute data in movici format.
+
+        :return: Dictionary with 'data' key and optionally 'row_ptr' for sparse arrays
+        """
+        result = {"data": self.data.to_array()}
+        if self.is_sparse:
+            result["row_ptr"] = self.indptr.to_array()
+        return result
+
+
+class UpdateAttribute(Base):
+    """Junction table linking updates to their attribute data."""
+
+    __tablename__ = "update_attribute"
+    __table_args__ = (UniqueConstraint("update_id", "attribute_data_id", name="uq_update_attr"),)
+
+    update_id = Column(Integer, ForeignKey("update.id", ondelete="CASCADE"), primary_key=True)
+    attribute_data_id = Column(
+        Integer, ForeignKey("attribute_data.id", ondelete="CASCADE"), primary_key=True
+    )
+
+
+class SimulationDatabase:
```

Comment:
It might be nice to also store the initial datasets in the database so that they're all together. Added benefit is that we then also have a snapshot of the datasets as they were during the simulation